src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs

```diff
@@ -219,6 +227,33 @@ private void Visit(Attachment attachment, string attachmentPointer)
             {
                 Visit(attachment.FileLocation, attachmentPointer.AtProperty(SarifPropertyName.FileLocation));
             }
+
+            if (attachment.Description != null)
+            {
+                Visit(attachment.Description, attachmentPointer.AtProperty(SarifPropertyName.Description));
+            }
+
+            if (attachment.Regions != null)
+            {
+                Region[] regions = attachment.Regions.ToArray();
```

Comment:
>egions.ToArray() [](start = 47, length = 16)

Why are you calling ToArray here and elsewhere? These collections are already indexable, aren't they? My concern is that you might be unnecessarily allocating memory. Can you either confirm that these calls aren't allocating or remove them in preference of indexing directly into the IList<Rectangle>? Comment applies here and elsewhere of course.

Reply:
I didn't know you could [index a list](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.item?view=netframework-4.7.2). I'll change it. But isn't it `O(n^2)` in time to iterate over a list that way? Or does .NET optimize storage of small lists to allow efficient random access?

---
In reply to: [239513818](https://github.com/Microsoft/sarif-sdk/pull/1149#discussion_r239513818) [](ancestors = 239513818)

Reply:
TIL, thanks Michael.

---
In reply to: [239524038](https://github.com/Microsoft/sarif-sdk/pull/1149#discussion_r239524038) [](ancestors = 239524038,239513818)

Reply:
btw - i did eventually locate the Linq ToArray source and can confirm this code allocates. So we'll save memory w/the change. 

---
In reply to: [239528433](https://github.com/Microsoft/sarif-sdk/pull/1149#discussion_r239528433) [](ancestors = 239528433,239524038,239513818)